### PR TITLE
feat(examples): add pipeline tool and pedestrian distance estimation demo

### PR DESCRIPTION
**Pull request by tpoisonooo**
增加人体距离估计 Demo

### CODE REVIEW

**Comment by tpoisonooo** [comment]
@BUG1989   please kindly review.

File: examples/CMakeLists.txt [hunk 1]
```diff
@@ -104,8 +104,11 @@ ENDIF()
 FIND_PACKAGE(OpenCV QUIET)
 
 IF (OpenCV_FOUND)
+    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread")
```
Comment (kalcohol) [inline examples/CMakeLists.txt:107]: pthread is not available for each platform. check source/CMakeLists.txt for more info.
source/CMakeLists.txt L290 ~ L299
```
# pthread
IF (NOT OHOS)
	INCLUDE (${PROJECT_SOURCE_DIR}/cmake/libraries/pthread.cmake)
	TENGINE_CHECK_LIB_PTHREAD (_has_pthread)
	IF (_has_pthread)
		TENGINE_USE_LIB_PTHREAD (${TENGINE_LITE_NAME}-static ON)
		TENGINE_USE_LIB_PTHREAD (${TENGINE_LITE_NAME}        OFF)
	ENDIF()
	UNSET (_has_pthread)
ENDIF()
```

File: examples/pipeline/.gitignore [hunk 2]
```diff
@@ -0,0 +1,8 @@
+*.o
+*.tmfile
+test_int
+test_camera_display
+test_preview_ped
+.vscode/
+mobilenet_ssd.tmfile
+libtengine-lite.so
```
Comment (kalcohol) [inline examples/pipeline/.gitignore:8]: is this file nesscessary for end user?

File: examples/pipeline/actor/draw_video.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      // fprintf(stdout, "show\n");
+      cv::imshow("camera", mat);
+      cv::waitKey(25);
+    }
+  }
+
+  ~DrawVideo() { cv::destroyAllWindows(); }
+};
+
+} // namespace pipe
```
Comment (kalcohol) [inline examples/pipeline/actor/draw_video.h:49]: a blank new line

File: examples/pipeline/actor/pedestrian_detection.h [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+  std::unique_ptr<float> m_input_data;
+
+  int img_h = 300;
+  int img_w = 300;
+  float mean[3] = {127.5f, 127.5f, 127.5f};
+  float scale[3] = {0.007843f, 0.007843f, 0.007843f};
+  float show_threshold = 0.5f;
+};
+
+} // namespace pipe
```
Comment (kalcohol) [inline examples/pipeline/actor/pedestrian_detection.h:236]: a blank new line

File: examples/pipeline/actor/spatial_distance_calculation.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+private:
+  std::deque<size_t> m_history_width;
+  const size_t MAX_LEN = 8;
+  const int WIDTH_THRESHOLD = 10;
+  const double SCALE = 100. / 0.27;
+};
+
+} // namespace pipe
```
Comment (kalcohol) [inline examples/pipeline/actor/spatial_distance_calculation.h:72]: a blank new line

File: examples/pipeline/actor/video_camera.h [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    m_opened = true;
+    cv::waitKey(1000 / std::max(1.0, rate));
+  }
+
+private:
+  std::string m_path;
+  bool m_opened;
+};
+
+} // namespace pipe
```
Comment (kalcohol) [inline examples/pipeline/actor/video_camera.h:89]: a blank new line

File: examples/pipeline/graph/graph.h [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    }
+  }
+
+private:
+  std::vector<std::unique_ptr<BaseNode>> m_nodes;
+  std::vector<std::unique_ptr<BaseEdge>> m_edges;
+  std::atomic<bool> m_running;
+  std::vector<std::thread> m_threads;
+};
+} // namespace pipe
```
Comment (kalcohol) [inline examples/pipeline/graph/graph.h:76]: a blank new line

File: examples/pipeline/test/test_camera_preview.cpp [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  auto draw = g.add_node<DrawVideo>();
+
+  auto cam_draw = g.add_edge<InstantEdge<cv::Mat>>(100);
+
+  cam->set_output<0>(cam_draw);
+  draw->set_input<0>(cam_draw);
+
+  g.start();
+  std::this_thread::sleep_for(std::chrono::milliseconds(60000));
+}
```
Comment (kalcohol) [inline examples/pipeline/test/test_camera_preview.cpp:43]: a blank new line

File: examples/pipeline/test/test_pedestrian_distance_estimation.cpp [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  cam->set_output<0>(cam_det);
+  detect_ped->set_input<0>(cam_det);
+  detect_ped->set_output<0>(det_dist);
+  dist_estimate->set_input<0>(det_dist);
+  dist_estimate->set_output<0>(dist_draw);
+  draw->set_input<0>(dist_draw);
+
+  g.start();
+  std::this_thread::sleep_for(std::chrono::milliseconds(600000));
+}
```
Comment (kalcohol) [inline examples/pipeline/test/test_pedestrian_distance_estimation.cpp:53]: a blank new line

File: examples/pipeline/actor/draw_video.h [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Copyright (c) 2021
+ * Author: tpoisonooo
+ */
+#pragma once
+#include "../graph/node.h"
```
Comment (kalcohol) [inline examples/pipeline/actor/draw_video.h:25]:
```suggestion
#include "pipeline/graph/node.h"
```
this usage may confuse user.
pls use "pipeline/graph/node.h" instead.

File: examples/pipeline/actor/draw_video.h [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+#include <opencv2/opencv.hpp>
```
Comment (kalcohol) [inline examples/pipeline/actor/draw_video.h:26]: opencv4 under consideration?

File: examples/pipeline/actor/draw_video.h [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    cv::Mat mat;
+    while (true) {
+
+      auto suc = input<0>()->pop(mat);
+      if (not suc) {
+        continue;
+      }
+      // fprintf(stdout, "show\n");
+      cv::imshow("camera", mat);
+      cv::waitKey(25);
```
Comment (kalcohol) [inline examples/pipeline/actor/draw_video.h:42]:
```suggestion
      cv::waitKey(1);
```
why 25ms?

File: examples/pipeline/actor/pedestrian_detection.h [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+  float score;
+};
+
+class PedestrianDetection : public Node<Param<cv::Mat>, Param<std::tuple<cv::Mat, cv::Rect>>> {
+public:
+  PedestrianDetection(std::string model_path) {
+    /* inital tengine */
+    init_tengine();
+    fprintf(stderr, "tengine-lite library version: %s\n",
+            get_tengine_version());
```
Comment (kalcohol) [inline examples/pipeline/actor/pedestrian_detection.h:50]: stdout is better.

File: examples/pipeline/actor/pedestrian_detection.h [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+    m_graph = create_graph(NULL, "tengine", model_path.c_str());
```
Comment (kalcohol) [inline examples/pipeline/actor/pedestrian_detection.h:51]: use context_t to adapt other backends
such as
```
m_ctx = create_context("ctx", 1);
auto ret = set_context_device("dev", m_dev, NULL, 0);
if (0 != ret)
{
    // just a warning
}
m_graph = create_graph(NULL, "tengine", model_path.c_str());
```

and m_dev should be inited firstly.

File: examples/pipeline/actor/pedestrian_detection.h [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+    if (m_graph == nullptr) {
+      fprintf(stderr, "create graph failed\n");
+      return;
+    }
+
+    /* set runtime options */
+    struct options opt;
+    opt.num_thread = 2;
```
Comment (kalcohol) [inline examples/pipeline/actor/pedestrian_detection.h:59]: thread should be configurable

File: examples/pipeline/actor/pedestrian_detection.h [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+    auto get_current_time = []() -> double {
+      struct timeval tv;
+      gettimeofday(&tv, NULL);
+      return tv.tv_sec * 1000.0 + tv.tv_usec / 1000.0;
+    };
+
+    /* prepare process input data, set the data mem to input tensor */
+    auto time1 = get_current_time();
+
+    cv::Mat input(img_h, img_w, CV_8UC3);
```
Comment (kalcohol) [inline examples/pipeline/actor/pedestrian_detection.h:160]: please consider using a cache

File: examples/pipeline/actor/pedestrian_detection.h [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+    cv::resize(mat, input, cv::Size(img_w, img_h));
+    cv::cvtColor(input, input, CV_BGR2RGB);
```
Comment (kalcohol) [inline examples/pipeline/actor/pedestrian_detection.h:162]: model may not need this color convert

File: examples/pipeline/actor/pedestrian_detection.h [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+    for (int h = 0; h < img_h; h++) {
+      for (int w = 0; w < img_w; w++) {
+        for (int c = 0; c < 3; c++) {
+          int in_index = h * img_w * 3 + w * 3 + c;
+          int out_index = c * img_h * img_w + h * img_w + w;
+          m_input_data.get()[out_index] =
+              (img_data[in_index] - mean[c]) * scale[c];
+        }
+      }
+    }
```
Comment (kalcohol) [inline examples/pipeline/actor/pedestrian_detection.h:176]: add a cache and combine this two steps will faster

File: examples/pipeline/actor/pedestrian_detection.h [hunk 19]
```diff
@@ -0,0 +1,236 @@
...
+
+private:
+  graph_t m_graph;
+  std::unique_ptr<float> m_input_data;
+
+  int img_h = 300;
+  int img_w = 300;
+  float mean[3] = {127.5f, 127.5f, 127.5f};
+  float scale[3] = {0.007843f, 0.007843f, 0.007843f};
+  float show_threshold = 0.5f;
```
Comment (kalcohol) [inline examples/pipeline/actor/pedestrian_detection.h:233]: should add a function to configuration these params

File: examples/pipeline/actor/spatial_distance_calculation.h [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Copyright (c) 2021
+ * Author: tpoisonooo
+ */
+#pragma once
+#include "../graph/node.h"
```
Comment (kalcohol) [inline examples/pipeline/actor/spatial_distance_calculation.h:25]:
```suggestion
#include "pipeline/graph/node.h"
```

File: examples/pipeline/actor/video_camera.h [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+ */
+
+/*
+ * Copyright (c) 2021
+ * Author: tpoisonooo
+ */
+#pragma once
+#include <thread>
+#include <mutex>
+#include "../graph/node.h"
```
Comment (kalcohol) [inline examples/pipeline/actor/video_camera.h:27]:
```suggestion
#include "pipeline/graph/node.h"
```

File: examples/pipeline/test/test_camera_preview.cpp [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+ * under the License.
+ */
+
+/*
+ * Copyright (c) 2021
+ * Author: tpoisonooo
+ */
+#include "../actor/draw_video.h"
+#include "../actor/video_camera.h"
+#include "../graph/graph.h"
```
Comment (kalcohol) [inline examples/pipeline/test/test_camera_preview.cpp:26]:
```suggestion
#include "pipeline/actor/draw_video.h"
#include "pipeline/actor/video_camera.h"
#include "pipeline/graph/graph.h"
```

File: examples/pipeline/test/test_pedestrian_distance_estimation.cpp [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+
+/*
+ * Copyright (c) 2021
+ * Author: tpoisonooo
+ */
+#include "../actor/draw_video.h"
+#include "../actor/pedestrian_detection.h"
+#include "../actor/video_camera.h"
+#include "../actor/spatial_distance_calculation.h"
+#include "../graph/graph.h"
```
Comment (kalcohol) [inline examples/pipeline/test/test_pedestrian_distance_estimation.cpp:28]:
```suggestion
#include "pipeline/actor/draw_video.h"
#include "pipeline/actor/pedestrian_detection.h"
#include "pipeline/actor/video_camera.h"
#include "pipeline/actor/spatial_distance_calculation.h"
#include "pipeline/graph/graph.h"
```

File: examples/pipeline/actor/pedestrian_detection.h (hunk 13 above)
Comment (tpoisonooo) [reply examples/pipeline/actor/pedestrian_detection.h:50]: `stdout` 就要引入`<iostream>`让 exe 变大 1-2M，要么简单的 `fprintf`；要么进一步用`spdlog`这种真正靠谱的库。  卡在中间用 `stdout`感觉并不好。

File: examples/pipeline/actor/pedestrian_detection.h (hunk 19 above)
Comment (tpoisonooo) [reply examples/pipeline/actor/pedestrian_detection.h:233]: 啊，本质上是需要整个  chain of preprocess 变 configurable.

Comment (tpoisonooo) [reply examples/pipeline/actor/pedestrian_detection.h:233]: 前处理、推理、后处理是绑定的。`img_h`、`img_w`和`show_threshold`配置是有意义的。

别的 `mean`、`scale`变  configurable 没有实际作用，注定是个脆弱的设计。

File: examples/pipeline/actor/draw_video.h (hunk 11 above)
Comment (tpoisonooo) [reply examples/pipeline/actor/draw_video.h:26]: 没用过  OCV4， 有别人写好的，再抄一个 `#include`

File: examples/pipeline/actor/pedestrian_detection.h (hunk 17 above)
Comment (tpoisonooo) [reply examples/pipeline/actor/pedestrian_detection.h:162]: 这个问题和前处理参数可配置一样：要不要把  pedestrian_detection.h 做成  det model 前后处理可配置的？

可以吧。晚点改好。

**Review by kalcohol — APPROVED** [review]
先提交，再改善吧。

File: examples/pipeline/actor/pedestrian_detection.h (hunk 17 above)
Comment (tpoisonooo) [reply examples/pipeline/actor/pedestrian_detection.h:162]: 在  https://github.com/OAID/Tengine/pull/928 改了。前后处理、模型都是外面 set 进去的。

**Pull request merged**